pkg/istio-agent/xds-proxy.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
+// This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
+// as the new connection may not go to the same istiod. Vice versa case also applies.
+func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	errChan := make(chan error)
+	requestsChan := make(chan *discovery.DiscoveryRequest)
+	responsesChan := make(chan *discovery.DiscoveryResponse)
+
+	proxyLog.Infof("connecting to upstream %s", p.istiodAddress)
+	upstreamConn, err := grpc.Dial(p.istiodAddress, p.istiodDialOptions...)
+	if err != nil {
+		proxyLog.Errorf("failed to connect to upstream %s: %v", p.istiodAddress, err)
+		return err
+	}
+
+	xds := discovery.NewAggregatedDiscoveryServiceClient(upstreamConn)
+	upstream, err := xds.StreamAggregatedResources(context.Background(),
+		grpc.MaxCallRecvMsgSize(defaultClientMaxReceiveMessageSize))
+	if err != nil {
+		proxyLog.Errorf("failed to create upstream grpc client: %v", err)
+		return err
+	}
+
+	go func() {
+		for {
+			// from istiod
+			resp, err := upstream.Recv()
+			if err != nil {
+				proxyLog.Errorf("upstream recv error: %v", err)
+				errChan <- err
+				return
+			}
+			responsesChan <- resp
+		}
+	}()
+	go func() {
+		for {
+			// From Envoy
+			req, err := downstream.Recv()
+			if err != nil {
+				proxyLog.Errorf("downstream recv error: %v", err)
+				errChan <- err
+				return
+			}
+			// forward to istiod
+			requestsChan <- req
+		}
+	}()
+
+	for {
+		select {
+		case err := <-errChan:
+			// error receiving from downstream envoy
+			// recycle connection.
+			_ = upstream.CloseSend()
+			// todo close downstream?
+			return err
+		case req := <-requestsChan:
+			if err = upstream.Send(req); err != nil {
+				proxyLog.Errorf("upstream send error: %v", err)
+				return err
+			}
+		case resp := <-responsesChan:
+			if err := downstream.Send(resp); err != nil {
+				proxyLog.Errorf("downstream send error: %v", err)
+				// we cannot return partial error and hope to restart just the downstream
+				// as we are blindly proxying req/responses. For now, the best course of action
+				// is to terminate upstream connection as well and restart afresh.
+				return err
+			}
+		case <-p.stopChan:
+			_ = upstream.CloseSend()
+			return nil
+		}
+	}
+}
+
+func (p *XdsProxy) DeltaAggregatedResources(server discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesServer) error {
+	return errors.New("delta XDS is not implemented")
+}
+
+func (p *XdsProxy) close() {
+	p.stopChan <- struct{}{}
+	if p.downstreamGrpcServer != nil {
+		_ = p.downstreamGrpcServer.Stop
+	}
+	if p.downstreamListener != nil {
+		_ = p.downstreamListener.Close()
+	}
+}
+
+// TODO reuse code from SDS
+// TODO reuse the connection, not just code
+func setUpUds(udsPath string) (net.Listener, error) {
```

Comment:
Why do we need uds path?

Reply:
from envoy to agent. this is a copy of sds code but needs to be refactored out to share common things, including sds functionality (cc @howardjohn ) . To keep the PR small, its doing things piece by piece.